### DISCUSSION TITLE
How to render text upside-down? - asked by @PPakalns

### DISCUSSION THREAD

**Discussion by Blatko1** [Blatko1/wgpu-text#25] [Q&A] [answered]
> I have similar issue and I am using patched wgpu_text to resolve it.
> 
> In my case y axis for world is in opposite direction of assumed y axis direction in wgpu_text. (I need to flip y axis of the text)
> 
> I have tried to resolve it by modifying matrix or using negative position coordinates, but it looks like to flip text (y axis) I need additional information about texture box size etc. Maybe I am doing something wrong.
> 
> But to resolve the issue I just modified shader to render text in needed direction.
> 
> [00194e8](https://github.com/Blatko1/wgpu-text/commit/00194e8643bc7cd9ac9dd60020c6f2b51434100a)

- asked by @PPakalns

**Comment by Blatko1** [comment, marked as answer]
As I understand, you want to make your text look flipped or, in other words, upside down. Did I understand correctly?
You can use that alternate version of the shader you linked, which turns every quad upside-down, but the **character alignment** will look a bit off - or you can use the *modified* version of the **orthogonal matrix**. I have tried experimenting a little with the orthogonal matrices by changing matrix element signs (minuses to pluses and the other way around) and have found that you can make your text look upside-down if you make your matrix look like this:
```rust
    [
        [-2.0 / view_width, 0.0,               0.0, 0.0],
        [0.0,               2.0 / view_height, 0.0, 0.0],
        [0.0,               0.0,               1.0, 0.0],
        [1.0,              -1.0,               0.0, 1.0]
    ]
```

Here's the default orthogonal matrix so you can compare the differences (notice the modified number signs):
```rust
    [
        [2.0 / view_width, 0.0,               0.0, 0.0],
        [0.0,             -2.0 / view_height, 0.0, 0.0],
        [0.0,              0.0,               1.0, 0.0],
        [-1.0,             1.0,               0.0, 1.0]
    ]
```

Afterwards, you get this result (you can modify the tex position in the code): 

![image](https://github.com/Blatko1/wgpu-text/assets/51339494/4cc48264-7e53-4b9e-ad6d-539b44459187)

**Reply by PPakalns** [reply]
Hi thank you for your answer and creating discussion for this problem, I was away from computer for a couple of days.

I will try your specific solution, but I already tried similar solutions as you provided, I can get the text to be rendered flipped upside down (In my case then it is rendered in correct direction), but the problem is that it doesn't stick to "specific" world position.  When i move up, down in world text moves in opposite direction. It is only rendered at the correct position when i vertically center the world position where the text is rendered and then text is rendered at vertical center of the screen. When i move up or down from this position, text moves in opposite direction. 

I concluded that i need to flip text (quad) vertices themself in opposite direction, that is what i am doing with the modified shader. 

Probably minimal example should include some kind of background coordinate axis, like rendered points at positions (0, 0), (0, 1), (1, 0).  Then the world could be moved around with text to showcase world axis directions and text stickiness.
